File: minimove/Assets/MiniGames/SafeCracker.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;

public enum SolveState {
	Searching,
	Found,
	Activated,
	Afterglow,
}

public class SafeCracker : MiniGame {
	private System.Random rnd = new System.Random ();
	private static readonly Color SafeColor = new Color (0.7f, 0.6f, 0f);
	int [] angles;
	int [] targetAngles;
	int [] lastClickAngles;
	int [] solvedLocks;
	SolveState [] solveState;
	int angleSteps = 12;
	int numberOfLocks = 2;
	int playersReady = 0;

	public SafeCracker(GameFlow gameFlow) : base(gameFlow) {
	}

	public override string StatusMessage() {
		string result = "";
		for (int i = 0; i < angles.Length; i++) {
			result += "[" + i + "] = " + angles [i] + ", ";
		}/*
		result += "\n";
		for (int i = 0; i < targetAngles.Length; i++) {
			result += "targetAngles[" + i + "] = " + targetAngles [i] + ", ";
		}
		result += "\n";
		for (int i = 0; i < solvedLocks.Length; i++) {
			result += "solvedLocks[" + i + "] = " + solvedLocks [i] + ", ";
		}*/

		return result;
	}

	public override void StartGame () {
		OnFinished onIntroBlinkingFinished = delegate() {
			playersReady++;

			if (playersReady == gameFlow.Players.Count) {
				angles = new int[gameFlow.Players.Count];
				foreach (var player in gameFlow.Players) {
					angles [player.PlayerNumber] = 0;
				}

				targetAngles = new int[gameFlow.Players.Count];
				foreach (var player in gameFlow.Players) {
					targetAngles [player.PlayerNumber] = rnd.Next (angleSteps) * 360 / angleSteps;
				}

				lastClickAngles = new int[gameFlow.Players.Count];
				foreach (var player in gameFlow.Players) {
					lastClickAngles [player.PlayerNumber] = 0;
				}

				solvedLocks = new int[gameFlow.Players.Count];
				foreach (var player in gameFlow.Players) {
					solvedLocks [player.PlayerNumber] = 0;
				}

				solveState = new SolveState[gameFlow.Players.Count];
				foreach (var player in gameFlow.Players) {
					solveState [player.PlayerNumber] = SolveState.Searching;
				}
			}
		};

		gameFlow.PlaySound("SafeAnnounceSound");

		foreach (var player in gameFlow.Players) {
			List<AnimationPart> parts = new List<AnimationPart> ();
			parts.Add(new AnimationPart(SafeColor*0.2f, 0.4f, 0.6f));
			parts.Add(new AnimationPart(SafeColor, 0.5f, 0.7f));
			parts.Add(new AnimationPart(Color.black, 0.4f));

			gameFlow.StartCoroutine(player.SphereColorAnimation (parts, onIntroBlinkingFinished));
		}
	}

	public IEnumerator ResetSearch(MovePlayer player, float delay) {
		yield return new WaitForSeconds (delay);

		solvedLocks [player.PlayerNumber]++;
		if (solvedLocks [player.PlayerNumber] == numberOfLocks) {
			gameFlow.endCurrentGame (player);
		} else {
			targetAngles [player.PlayerNumber] = rnd.Next (angleSteps) * 360 / angleSteps;
			solveState [player.PlayerNumber] = SolveState.Searching;
		}
	}

	public override void Update () {
		if (playersReady < gameFlow.Players.Count) {
			return;
		}

		int threshold = 360 / angleSteps;
		foreach (var player in gameFlow.Players) {
			int newAngle = (int)player.SafeAngle ();
			if (Mathf.Abs(newAngle - lastClickAngles [player.PlayerNumber]) > 20f ) {
				gameFlow.PlaySound ("SafeClickSound", 0.3f);
				lastClickAngles [player.PlayerNumber] = newAngle;
			}
			angles [player.PlayerNumber] = newAngle;

			bool isNear = (angles [player.PlayerNumber] >= targetAngles [player.PlayerNumber] &&
			               angles [player.PlayerNumber] <= targetAngles [player.PlayerNumber] + threshold);

			// State transitions
			switch (solveState [player.PlayerNumber]) {
				case SolveState.Searching:
					if (isNear && !player.IsTriggerPressed()) {
						solveState [player.PlayerNumber] = SolveState.Found;
					}
					break;
				case SolveState.Found:
					if (!isNear) {
						solveState [player.PlayerNumber] = SolveState.Searching;
					} else if (player.IsTriggerPressed ()) {
						solveState [player.PlayerNumber] = SolveState.Activated;
					}
					break;
				case SolveState.Activated:
					gameFlow.StartCoroutine (ResetSearch (player, 0.2f));
					solveState [player.PlayerNumber] = SolveState.Afterglow;
					break;
				case SolveState.Afterglow:
					break;
			}

			// State actions
			switch (solveState [player.PlayerNumber]) {
				case SolveState.Searching:
					player.LEDColor = Color.black;
					player.Rumble = 0f;
					break;
				case SolveState.Found:
					player.LEDColor = Color.black;
					player.Rumble = 0.5f;
					break;
				case SolveState.Activated:
					player.LEDColor = SafeColor;
					player.Rumble = 0.5f;
					break;
				case SolveState.Afterglow:
					player.LEDColor = SafeColor;
					player.Rumble = 1f;
					break;
			}

			/*
			if () {
				player.Rumble = 0.5f;
				if (player.IsTriggerPressed ()) {
					player.LEDColor = SafeColor;
				} else {
					player.LEDColor = Color.black;
				}
			} else {
				player.LEDColor = Color.black;
				player.Rumble = 0f;
			}*/
		}
	}

	public override bool CanSupportPlayers(int players) {
		return true;
	}
}

File: python/minigames/safecracker.py
#
# Parley Who Vertigo
# Copyright 2016, 2017 Thomas Perl, Josef Who
#
# Permission to use, copy, modify, and/or distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
# REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND
# FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
# INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
# LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
# OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
# PERFORMANCE OF THIS SOFTWARE.
#


import minigame
import random
import moveplayer
import coroutine

from color import Color


class SolveState(object):
    Searching, Found, Activated, Afterglow = range(4)


class SafeCracker(minigame.MiniGame):
    ENABLED = True

    SAFE_COLOR = Color(0.91, 0.78, 0.0)

    def __init__(self, gameflow):
        super().__init__(gameflow)
        self.angles = []
        self.target_angles = []
        self.last_click_angles = []
        self.solved_locks = []
        self.solve_state = []
        self.angle_steps = 12
        self.number_of_locks = 2
        self.players_ready = 0

    def status_message(self):
        return 'You crack me up lil\' buddy'

    def base_color(self):
        return self.SAFE_COLOR

    def start_game(self):
        def on_intro_blinking_finished():
            self.players_ready += 1

            if self.players_ready == len(self.gameflow.players):
                self.angles = [0 for player in self.gameflow.players]

                self.target_angles = [random.randint(0, self.angle_steps-1) * 360 / self.angle_steps
                                      for player in self.gameflow.players]

                self.last_click_angles = [0 for player in self.gameflow.players]
                self.solved_locks = [0 for player in self.gameflow.players]
                self.solve_state = [SolveState.Searching for player in self.gameflow.players]

        self.gameflow.play_sound('SafeAnnounceSound')

        for player in self.gameflow.players:
            parts = []
            parts.append(moveplayer.AnimationPart(self.SAFE_COLOR * 0.2, 0.4, 0.6))
            parts.append(moveplayer.AnimationPart(self.SAFE_COLOR, 0.5, 0.7))
            parts.append(moveplayer.AnimationPart(Color.BLACK, 0.4))

            self.gameflow.start_coroutine(player.sphere_color_animation(parts, on_intro_blinking_finished))

    def reset_search(self, player, delay):
        yield coroutine.WaitForSeconds(delay)

        self.solved_locks[player.player_number] += 1
        if self.solved_locks[player.player_number] == self.number_of_locks:
            self.gameflow.end_current_game(player)
        else:
            self.target_angles[player.player_number] = random.randint(0, self.angle_steps-1) * 360 / self.angle_steps
            self.solve_state[player.player_number] = SolveState.Searching

    def update(self):
        tunables = self.gameflow.get_tunables()

        if self.players_ready < len(self.gameflow.players):
            return

        threshold = int(360 / self.angle_steps)

        for player in self.gameflow.players:
            new_angle = int(player.safe_angle())
            if abs(new_angle - self.last_click_angles[player.player_number]) > 20.0:
                self.gameflow.play_sound('SafeClickSound', 0.3)
                self.last_click_angles[player.player_number] = new_angle
            self.angles[player.player_number] = new_angle

            is_near = (self.angles[player.player_number] >= self.target_angles[player.player_number] and
                       self.angles[player.player_number] <= self.target_angles[player.player_number] + threshold)

            # State transitions
            if self.solve_state[player.player_number] == SolveState.Searching:
                if is_near and not player.is_trigger_pressed():
                    self.solve_state[player.player_number] = SolveState.Found
            elif self.solve_state[player.player_number] == SolveState.Found:
                if not is_near:
                    self.solve_state[player.player_number] = SolveState.Searching
                elif player.is_trigger_pressed():
                    self.solve_state[player.player_number] = SolveState.Activated
            elif self.solve_state[player.player_number] == SolveState.Activated:
                self.gameflow.start_coroutine(self.reset_search(player, 0.2))
                self.solve_state[player.player_number] = SolveState.Afterglow
            elif self.solve_state[player.player_number] == SolveState.Afterglow:
                ...

            # State actions
            if self.solve_state[player.player_number] == SolveState.Searching:
                player.led_color = self.SAFE_COLOR * tunables.ColorIntensityDuringGameplay
                player.rumble = 0.0
            elif self.solve_state[player.player_number] == SolveState.Found:
                player.led_color = self.SAFE_COLOR * tunables.ColorIntensityDuringGameplay
                player.rumble = 0.5
            elif self.solve_state[player.player_number] == SolveState.Activated:
                player.led_color = self.SAFE_COLOR
                player.rumble = 0.5
            elif self.solve_state[player.player_number] == SolveState.Afterglow:
                player.led_color = self.SAFE_COLOR
                player.rumble = 1.0
